--- /home/szbaijie/acho/mcp251xfd/mcp251xfd/mcp251xfd.h	2022-09-29 14:18:28.256177681 +0300
+++ /home/szbaijie/acho/mcp251xfd/mcp251xfd/mcp251xfd.h.patched	2022-09-29 14:23:33.988402900 +0300
@@ -19,6 +19,34 @@
 #include <linux/regulator/consumer.h>
 #include <linux/spi/spi.h>
 
+/* TID */
+/* ACHO */
+
+//#if (LINUX_VERSION_CODE < KERNEL_VERSION(4,16,0))
+#define sizeof_field(T, m) (sizeof(((T *)0)->m))
+//#endif
+/* TODO CHECK VERSION */
+/*
+*/
+
+extern void __compiletime_error("bad bitfield mask")
+__bad_mask(void);
+
+static __always_inline u64 field_multiplier(u64 field)
+{
+	if ((field | (field - 1)) & ((field | (field - 1)) + 1))
+		__bad_mask();
+	return field & -field;
+}
+
+static __always_inline u64 field_mask(u64 field)
+{
+	return field / field_multiplier(field);
+}
+
+/* TID */
+/* ACHO */
+
 /* MPC251x registers */
 
 /* CAN FD Controller Module SFR */
